ports/raspberrypi/common-hal/socketpool/Socket.c

```diff
@@ -745,123 +746,150 @@ socketpool_socket_obj_t *common_hal_socketpool_socket(socketpool_socketpool_obj_
     return socket;
 }
 
-int socketpool_socket_accept(socketpool_socket_obj_t *self, uint8_t *ip, uint32_t *port) {
-    return -MP_EBADF;
-}
-
-socketpool_socket_obj_t *common_hal_socketpool_socket_accept(socketpool_socket_obj_t *socket,
-    uint8_t *ip, uint32_t *port) {
-    if (socket->type != MOD_NETWORK_SOCK_STREAM) {
-        mp_raise_OSError(MP_EOPNOTSUPP);
+int socketpool_socket_accept(socketpool_socket_obj_t *self, uint8_t *ip, uint32_t *port, socketpool_socket_obj_t *accepted) {
+    if (self->type != MOD_NETWORK_SOCK_STREAM) {
+        return -MP_EOPNOTSUPP;
     }
 
-    // Create new socket object, do it here because we must not raise an out-of-memory
-    // exception when the LWIP concurrency lock is held
-    socketpool_socket_obj_t *socket2 = m_new_ll_obj_with_finaliser(socketpool_socket_obj_t);
-    socket2->base.type = &socketpool_socket_type;
+    if (common_hal_socketpool_socket_get_closed(self)) {
+        return -MP_EBADF;
+    }
 
     MICROPY_PY_LWIP_ENTER
 
-    if (socket->pcb.tcp == NULL) {
+    if (self->pcb.tcp == NULL) {
         MICROPY_PY_LWIP_EXIT
-        m_del_obj(socketpool_socket_obj_t, socket2);
```

Comment:
safe to remove this m_del_obj, it's just an optimization. gc will clean it up later if it was freshly allocated.

Reply:
It is still below. (New line 847) This diff isn't great because most of the logic here was kept the same even though it was moved to a non-allocating version of accept that the allocating version uses.